src/common/options/global.yaml.in

```diff
@@ -301,6 +301,28 @@ options:
   - startup
   # default changed by common_preinit()
   with_legacy: true
+- name: finisher_batch_events
+  type: uint
+  level: advanced
+  desc: how many events can finisher process continuously before giving up CPU to allow other threads to run
+  default: 5
+  services:
+  - common
+  see_also:
+  - finisher_sleep_interval
+  flags:
+  - runtime
+- name: finisher_sleep_interval
+  type: uint
+  level: advanced
+  desc: how long will the finisher thread sleep when giving up CPU, in nanoseconds
+  default: 5
```

Comment:
is this configuration useful anywhere outside of the mds? maybe instead of global configuration, it should be passed into `Finisher`'s constructor at runtime? that way only the ceph-mds finisher threads would have to sleep

Reply:
> is this configuration useful anywhere outside of the mds?

In theory, this should happen not only in MDS. 

>  maybe instead of global configuration, it should be passed into `Finisher`'s constructor at runtime? that way only the ceph-mds finisher threads would have to sleep

In the Finisher's constructor we cannot tune or change it if needed. Just in my env and test, sleeping for 5ns could make sure the Finisher thread will be scheduled out and won't sheduled back immediately. And also 1ns will also work well till now of my tests, I just saw a similar issue once, from the log it should be another issue, which I am still debugging it. So 1ns should be work well too.

IMO, allowing it in runtime could give us a chance to tune it if the default 5ns will introduce perf issue in some corner cases in future ?

Reply:
And also this could be set to 1 nanoseconds instead.

Reply:
@lxbsz how about implement `mds_lock` using a FIFO mutex, like https://gist.github.com/tchaikov/0b04e87c539732519ed164bf3af58e05 ?

sure, there is more work to do, like

- make `SafeTimer` accept `fair_mutex`
- change all the places where `ceph::mutex` is used to pass `mds_lock`
- implement `is_locked()` method

but i think it'd be better to have a sleep call in `Finisher`.

Reply:
Yeah, I was also thinking of this. Just not sure whether will this cause other performance issue ?

For Finisher thread in mds, the fair mutex unlock() will always let the finisher thread to be scheduled out even no waiter of this lock, then the finisher thread will be scheduled back immediately since the finisher thread will acquire/release the `mds_lock` frequently.

Won't will fair lock introduce more bubbles ?

Reply:
It seems your fair mutex lock above won't have this issue...

Reply:
> the fair mutex unlock() will always let the finisher thread to be scheduled out even no waiter of this lock

IMO, on linux the POSIX mutex is always implemented using futex. so on the the "fast" scenario, the caller is not swapped out.

Reply:
Fair mutex makes sense to me.

Reply:
@tchaikov Will you push your fair lock code beforehand ? Or together with this fixing ?

@batrick As I remembered last year we were hitting the same issue in fuse client code. But I couldn't remember the exact issue or PR in which I have mentioned this.

Reply:
#42556

Reply:
Have fixed this in new PR in https://github.com/ceph/ceph/pull/42620 by using the fair mutex. Will close this one.